### PR TITLE
fix: dependency target json parsing issue of old SPM `Package.swift`

### LINKED ISSUE #299: Gazelle fails to parse SPM dependency target with Nimble
## Symptom

I was trying to run `bazel run tidy` to generate build files with unit test codes which is written with `Quick` and `Nimble` test frameworks. And found the error below

```
INFO: Running command line: bazel-bin/swift_update_pkgs
gazelle: index 1 expected to be type 'string', but was type '<nil>'
```

There is a sample test in the forked repo here
https://github.com/cgrindel/swift_bazel/compare/main...tingyao-ui:issue/tingyao-nimble-example?expand=1

## Root Cause

`gazelle` plugin returns an error from `gazelle/internal/spdump/target_dependancy.go#L38: func (pr *ProductReference) UnmarshalJSON(b []byte) error`
```golang
func (pr *ProductReference) UnmarshalJSON(b []byte) error {
	var err error
	var raw []any
	if err = json.Unmarshal(b, &raw); err != nil {
		return err
	}
	if pr.ProductName, err = jsonutils.StringAtIndex(raw, 0); err != nil {
		return err
	}
// pr.DependencyName should not be nil
	if pr.DependencyName, err = jsonutils.StringAtIndex(raw, 1); err != nil {
		return err
	}
	return nil
}

```

Please refer to the screenshot below.

CleanShot 2023-03-23 at 12 16 18@2x

`Nimble` has a dependency on `CwlPreconditionTesting`.
`CwlPreconditionTesting` has a dependency target to `CwlCatchException` and its `Package.swift` is in the legacy format of the swift 5.0 version.
So that the JSON output of `swift package dump-package` won't have a correct dependency name

```json
## ➜  CwlPreconditionTesting git:(2.1.0) swift package dump-package   
{
  "cLanguageStandard" : null,
  "cxxLanguageStandard" : null,
  "dependencies" : [
    {
      "sourceControl" : [
        {
          "identity" : "cwlcatchexception",
          "location" : {
            "remote" : [
              "https://github.com/mattgallagher/CwlCatchException.git"
            ]
          },
          "productFilter" : null,
          "requirement" : {
            "range" : [
              {
                "lowerBound" : "2.0.0",
                "upperBound" : "3.0.0"
              }
            ]
          }
        }
      ]
    }
  ],
  "name" : "CwlPreconditionTesting",
  "packageKind" : {
    "root" : [
      "/Users/erichsu/Developer/swift_bazel/examples/nimble_example/.build/checkouts/CwlPreconditionTesting"
    ]
  },
  "pkgConfig" : null,
  "platforms" : [

  ],
  "products" : [
    {
      "name" : "CwlPreconditionTesting",
      "settings" : [

      ],
      "targets" : [
        "CwlPreconditionTesting",
        "CwlMachBadInstructionHandler"
      ],
      "type" : {
        "library" : [
          "automatic"
        ]
      }
    },
    {
      "name" : "CwlPosixPreconditionTesting",
      "settings" : [

      ],
      "targets" : [
        "CwlPosixPreconditionTesting"
      ],
      "type" : {
        "library" : [
          "automatic"
        ]
      }
    }
  ],
  "providers" : null,
  "swiftLanguageVersions" : null,
  "targets" : [
    {
      "dependencies" : [
        {
          "target" : [
            "CwlMachBadInstructionHandler",
            null
          ]
        },
        {
          "product" : [
            "CwlCatchException",
            null, # Parsing error fails here
            null,
            null
          ]
        }
      ],
      "exclude" : [

      ],
      "name" : "CwlPreconditionTesting",
      "resources" : [

      ],
      "settings" : [

      ],
      "type" : "regular"
    },
    {
      "dependencies" : [

      ],
      "exclude" : [

      ],
      "name" : "CwlPosixPreconditionTesting",
      "resources" : [

      ],
      "settings" : [

      ],
      "type" : "regular"
    },
    {
      "dependencies" : [

      ],
      "exclude" : [

      ],
      "name" : "CwlMachBadInstructionHandler",
      "resources" : [

      ],
      "settings" : [

      ],
      "type" : "regular"
    },
    {
      "dependencies" : [
        {
          "byName" : [
            "CwlPreconditionTesting",
            null
          ]
        },
        {
          "byName" : [
            "CwlPosixPreconditionTesting",
            null
          ]
        }
      ],
      "exclude" : [

      ],
      "name" : "CwlPreconditionTestingTests",
      "resources" : [

      ],
      "settings" : [

      ],
      "type" : "test"
    }
  ],
  "toolsVersion" : {
    "_version" : "5.0.0"
  }
}
```

### PR DESCRIPTION
**Pull request by tingyao-ui**
## Summary

For Issue #299 
This is a workaround to utilize the product name as a dependency name from the target of `Package.swift` if the name is nil

## Implement Details

- Update gazelle plugins of `UnmarshalJSON`
- Update starlark `pkginfos` parsing logic

Closes #299

### CODE REVIEW

File: gazelle/internal/spdump/target_dependency.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ func (pr *ProductReference) UnmarshalJSON(b []byte) error {
 		return err
 	}
 	if pr.DependencyName, err = jsonutils.StringAtIndex(raw, 1); err != nil {
-		return err
+		pr.DependencyName = pr.ProductName
```
Comment (cgrindel) [inline gazelle/internal/spdump/target_dependency.go:48]: I went looking for a reference in the SPM code for how they interpret the single name.  I found [this](https://github.com/apple/swift-package-manager/blob/main/Sources/PackageModel/Target.swift#L59-L62).

>// this is hit only if this product is referenced `.byName(name)`
// which assumes the name of this product, its package, and its target
// all have the same name

So, I think that your change looks good to me.

Comment (cgrindel) [inline gazelle/internal/spdump/target_dependency.go:48]:
```suggestion
                // Per SPM code, a single name implies that the product, 
                // package, and target all have the same name.
                pr.DependencyName = pr.ProductName
```

File: swiftpkg/internal/pkginfos.bzl [hunk 2]
```diff
@@ -166,7 +166,7 @@ def _new_target_dependency_from_dump_json_map(dump_map):
     if product_list:
         product = _new_product_reference(
             product_name = product_list[0],
-            dep_name = product_list[1],
+            dep_name = product_list[1] or product_list[0],
```
Comment (cgrindel) [inline swiftpkg/internal/pkginfos.bzl:169]:
```suggestion
            // Per SPM code, a single name implies that the product, 
            // package, and target all have the same name.
            dep_name = product_list[1] or product_list[0],
```

**Review by cgrindel — APPROVED** [review]

**Comment by cgrindel** [comment]
It looks like you need to run `bazel run //:tidy` and check-in the updated files (e.g., `.bazelrc`).

File: gazelle/internal/spdump/target_dependency.go (hunk 1 above)
Comment (tingyao-ui) [reply gazelle/internal/spdump/target_dependency.go:50]: wow, nice to know that.

**Comment by tingyao-ui** [comment]
> It looks like you need to run `bazel run //:tidy` and check-in the updated files (e.g., `.bazelrc`).

got it, all updated, thanks

**Review by cgrindel — APPROVED** [review]
Thanks, again, for the pull request!

**Pull request merged**